[APM] Agent remote configuration

ogupte opened the pull request:
## Summary

Closes #34990. Adds support for remote configuration of agents within the APM UI.

This is based off the work of this PR: https://github.com/elastic/kibana/pull/36031.

![agent-remote-config-ui-3](https://user-images.githubusercontent.com/1967266/60377625-6dbda000-99cd-11e9-8c76-3bd1098b5d9f.gif)

sorenlouv committed:
[APM] Central configuration management

ogupte commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/save_configuration.ts:
```diff
@@ -0,0 +1,36 @@
+  const { client, config } = setup;
+
+  const params = {
+    type: 'index_or_alias', // TODO set correct type
```
I'm not sure the correct type that should be set here.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3687/)

simitt commented:
@ogupte I tested APM Server playing together with this branch. I did find some unexpected behavior related to `service.environment`. Entries in ES that only contain a `service.name` but no `service.environment` should not be returned if the query only contains `service.name`. Same thing the other way around, if the query contains `service.name` and `service.environment` no results that are not matching the `service.environment` part should be returned. During my tests entries that should not match based on the environment were returned in the result. Do you have any tests for that?

ogupte commented:
> @ogupte I tested APM Server playing together with this branch. I did find some unexpected behavior related to `service.environment`. Entries in ES that only contain a `service.name` but no `service.environment` should not be returned if the query only contains `service.name`. Same thing the other way around, if the query contains `service.name` and `service.environment` no results that are not matching the `service.environment` part should be returned. During my tests entries that should not match based on the environment were returned in the result. Do you have any tests for that?

I'll make sure to address this and add tests to make sure only entries with both `service.name` and `service.environment` are returned. Thank you for checking this.

ogupte committed:
[APM] Agent remote configuration UI
  - Creates apmCm index mapping on plugin initialization.
  - Links from home page to settings page
  - Polishes settings page to match design

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3944/)

formgeist committed:
Copy edits on ListSettings page

formgeist committed:
Updated snapshot

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4131/)

ogupte committed:
- Updates setting config flyout to match design.
- Adds ability to update/edit existing setting value
- Adds delete functionality to edit settings flyout

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4217/)

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -155,6 +181,42 @@ export function AddSettingFlyoutBody({ onSubmit }: { onSubmit: () => void }) {
+
         <EuiFormRow>
           {/* TODO: Move to AddSettingFlyout EuiFlyoutFooter instead */}
           <EuiButton
```
I guess this is replaced by the submit button in the `EuiFlyoutFooter` in `AddSettingFlyout`?

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -143,7 +169,7 @@ export function AddSettingFlyoutBody({ onSubmit }: { onSubmit: () => void }) {
             min={0}
             max={1}
-            step={0.1}
+            step={0.001}
```
I think this might need to change to `0.01` as I don't think we need to support 0.5%. @graphaelli thoughts?

formgeist replied:
Notes from Slack conversation: we agreed to keep the 0.001 validation because it should be possible for users to do 0.1% sample rate configuration.

formgeist reviewed (changes requested):
@ogupte Starting to take shape and it's looking really good. I have a few nits and changes to request.

formgeist commented:
I forgot to add this in as a note yesterday after the APM weekly, but there's a couple of larger things to address;

**"All environments" option**

Perhaps because the Opbeans apps don't have multiple environments, I haven't been able to test out the functionality in which a user is able to select "all environments" option which will bulk create configurations for each environment for the selected service with the set sample rate. Remind me, is this possible in this current implementation?

**Minor design feedbacks**

- The services list is currently not sorted alphabetically.
<img width="411" alt="Screenshot 2019-06-27 at 08 48 31" src="https://user-images.githubusercontent.com/4104278/60243364-5c36a400-98b8-11e9-892d-ca7db3e52799.png">
- The Flyout should not pre-fill the name and environment selects upon loading. In some cases, the user will have created a configuration for the preselected values and that's where they see an error.
![Kapture 2019-06-27 at 8 51 14](https://user-images.githubusercontent.com/4104278/60243551-c4858580-98b8-11e9-9725-e81b5f4fb9a4.gif)
- The sample rate input field could use some stricter input validation. I understand that `00000000` results in `0`, but perhaps having it restricted to max. 4 chars (because `x.xx` is the max amount of decimals we allow).

ogupte commented:
@formgeist  
> **"All environments" option**
> 
> Perhaps because the Opbeans apps don't have multiple environments, I haven't been able to test out the functionality in which a user is able to select "all environments" option which will bulk create configurations for each environment for the selected service with the set sample rate. Remind me, is this possible in this current implementation?

We could do something like this, but it would be a shortcut to adding one configuration per each **existing** environment. This means if a new environment comes only after this confuration is set, it would not apply to the new environment. I'm not sure if users will understand this subtly. Also, would selecting 'all environments' like this apply to those services with no value in `service.environemnt`?

formgeist commented:
@ogupte I can see how it might be confusing to users that 1. is it going to work for my future environments and 2. how does it apply to services where environment is not set. I reckon we leave it for now. It's more of a shortcut or making that bulk action easier for the user. Let's focus on getting this and other planned features in good shape for next week 👍

formgeist committed:
Updated naming

formgeist committed:
Updated help texts in Flyout

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/ListSettings.tsx:
```diff
@@ -0,0 +1,196 @@
+          <p>
+            We're excited to bring you a first look at how you can control your
+            services by fine-tuning agent configurations directly from Kibana.
+            Learn more about this new feature.
```
@bmorelli25 We're going to need a URL for the documentation page we're intending to set up for APM Agent Configuration.

bmorelli25 replied:
Let's go with: `https://www.elastic.co/guide/en/kibana/current/agent-configuration.html`

formgeist replied:
Added this in (cc @ogupte) 👍

bmorelli25 commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,96 @@
+              This is where you define the configuration you want to sync to
+              your service and agent. This initial version is supported by the
+              Elastic APM Java agent and for sample rate configuration only. We
+              will continue to add support for central configuration for other
```
This sentence needs to be updated to remove "central"

```suggestion
              will continue to add support for Agent configuration for other
```

ogupte committed:
Update x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx

sentence updated to remove "central"

Co-Authored-By: Brandon Morelli <bmorelli25@gmail.com>

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4550/)

formgeist committed:
Updated tooltip and callout content

formgeist committed:
Converted beta label and desc to callout

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -52,18 +64,25 @@ export function AddSettingsFlyout({ onClose, isOpen, onSubmit }: Props) {
         <EuiFlyoutBody>
           <EuiText>
             <p>
-              Please note only sample rate configuration is supported in this
```
@ogupte Fixed this as part of fixing some more copy

formgeist commented:
@bmorelli25 I'm not sure if you already did a pass on the copy, but I would like some help on possible improvements. Thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4583/)

bmorelli25 reviewed (commented):
@formgeist - looks like you updated the copy since yesterday. I'm liking the changes! Just a few recommendations.

bmorelli25 commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,313 @@
+
+        <EuiFormRow
+          label="Transaction sample rate"
+          helpText="Choose a rate between 0.00 and 1.0. Default configuration is 1.0 (100& of traces)."
```
```suggestion
          helpText="Choose a rate between 0.00 and 1.0. Default configuration is 1.0 (100% of traces)."
```

bmorelli25 commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,313 @@
+              <p>
+                If you wish to delete this configuration, please be aware that
+                the agents will continue to use the existing configuration until
+                it syncs with the APM Server.
```
```suggestion
                they sync with the APM Server.
```

ogupte committed:
Update x-pack/legacy/plugins/apm/public/components/app/Settings/ListSettings.tsx

fix wording

Co-Authored-By: Brandon Morelli <bmorelli25@gmail.com>

ogupte committed:
Update x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx

fix typo

Co-Authored-By: Brandon Morelli <bmorelli25@gmail.com>

ogupte committed:
Update x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx

fix wording

Co-Authored-By: Brandon Morelli <bmorelli25@gmail.com>

ogupte commented on x-pack/legacy/plugins/apm/public/components/app/Settings/ListSettings.tsx:
```diff
@@ -0,0 +1,198 @@
+          <EuiFlexItem grow={false}>
+            <EuiBetaBadge
+              label="Beta"
+              tooltipContent="This feature is still in development. If you have feedback, please reach out in our Discuss forum."
```
Unfortunately, you can't mouseover the BetaBadge tooltip to click on it.

bmorelli25 replied:
Ah, good to know. Thanks!

ogupte committed:
Deduplicate copy between callout and empty state message

Co-Authored-By: Brandon Morelli <bmorelli25@gmail.com>

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4758/)

ogupte committed:
- moves save config logic up one level so it can use button in footer
- adds last updated to config table
- fixes validation for all form fields
- adds text for toast notifications (create,edit,delete)
- makes service name clickable to open edit flyout
- adds i18n for all text

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4855/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4860/)

formgeist commented:
@bmorelli25 Thanks for taking a look. LGTM 🚢

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,346 @@
+        )
+      });
+    } else {
+      await saveCMConfiguration(configuration);
```
nit: createCMConfiguration

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,346 @@
+        text: (
+          <FormattedMessage
+            id="xpack.apm.settings.cm.editConfigFailedText"
+            defaultMessage="Something when wrong when editing the configuration for {serviceName}. Error: {errorMessage}"
```
`Something when wrong` > `Something went wrong`

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,346 @@
+        text: (
+          <FormattedMessage
+            id="xpack.apm.settings.cm.createConfigFailedText"
+            defaultMessage="Something when wrong when creating a configuration for {serviceName}. Error: {errorMessage}"
```
`Something when wrong` > `Something went wrong`

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,273 @@
+              defaultMessage: 'Sample rate must be between 0.00 and 1'
+            }
+          )}
+          isInvalid={
```
Looks like we have validation in two places (this component, and its parent). Would it make sense to have it one place?

dgieselaar commented on x-pack/legacy/plugins/apm/public/services/rest/apm/settings.ts:
```diff
@@ -0,0 +1,63 @@
+  });
+}
+
+export async function saveCMConfiguration(
```
nit: createCMConfiguration

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/helpers/input_validation.ts:
```diff
@@ -4,12 +4,14 @@
 
-export const withDefaultValidators = (validators = {}) => {
+export const withDefaultValidators = (
+  validators: { [key: string]: Schema } = {}
```
🤟🏻

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/get_environments/get_all_environments.ts:
```diff
@@ -0,0 +1,58 @@
+        environments: {
+          terms: {
+            field: SERVICE_ENVIRONMENT,
+            missing: 'ENVIRONMENT_NOT_SET',
```
should this be a shared const? Looks like we export it in various places.

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/get_environments/get_all_environments.ts:
```diff
@@ -0,0 +1,58 @@
+
+  const resp = await client.search(params);
+  const aggs = resp.aggregations;
+  const buckets = idx(aggs, _ => _.environments.buckets) || [];
```
do we need `idx` here? if so, should we update the type defs for the aggregations?

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/get_environments/get_all_environments.ts:
```diff
@@ -0,0 +1,58 @@
+  SERVICE_ENVIRONMENT
+} from '../../../../../common/elasticsearch_fieldnames';
+
+export async function getAllEnvionments({
```
`getAllEnvionments` > `getAllEnvironments`

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/get_environments/get_unavailable_environments.ts:
```diff
@@ -0,0 +1,48 @@
+  SERVICE_ENVIRONMENT
+} from '../../../../../common/elasticsearch_fieldnames';
+
+export async function getUnavailableEnvionments({
```
`getAllEnvionments` > `getAllEnvironments`

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/list_configurations.ts:
```diff
@@ -0,0 +1,29 @@
+    index: config.get<string>('apm_oss.cmIndex')
+  };
+
+  // const resp = await client<CentralConfiguration>('search', params);
```
We can remove this right?

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/save_configuration.ts:
```diff
@@ -0,0 +1,36 @@
+export type CMSaveConfigurationAPIResponse = PromiseReturnType<
+  typeof saveConfiguration
+>;
+export async function saveConfiguration({
```
nit: `createConfiguration`

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/save_configuration.ts:
```diff
@@ -0,0 +1,36 @@
+  configuration: CentralConfigurationIntake;
+  setup: Setup;
+}) {
+  const { client, config } = setup;
```
should this endpoint not do its own validation of whether a configuration exists for a given serviceName/environment? Seems like a stale UI could fairly easily allow the user to create one.

ogupte replied:
I can handle this in a different PR https://github.com/elastic/kibana/issues/40114

dgieselaar commented on x-pack/legacy/plugins/apm/server/routes/settings.ts:
```diff
@@ -0,0 +1,199 @@
+  throw Boom.boomify(err, { statusCode: 400 });
+};
+
+export function initSettingsApi(core: InternalCoreSetup) {
```
Looks like these are all fairly light on request validation. Do we need a little more security here perhaps?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Main/route_config/index.tsx:
```diff
@@ -47,6 +48,15 @@ export const routes: BreadcrumbRoute[] = [
+  {
+    exact: true,
+    path: '/settings',
+    component: ListSettings,
```
nit: `ListSettings` >  `SettingsList`. (ListSettings sounds like either an instruction rather than a component, or a component that displays settings for a list).

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Main/route_config/route_names.tsx:
```diff
@@ -4,6 +4,11 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+/*
```
dupe?

dgieselaar commented:
Nice work @ogupte! Just some minor things from my side.

bmorelli25 commented on x-pack/legacy/plugins/apm/public/components/app/Settings/ListSettings.tsx:
```diff
@@ -0,0 +1,290 @@
+      name: i18n.translate(
+        'xpack.apm.settings.cm.configTable.serviceNameColumnLabel',
+        {
+          defaultMessage: 'Service name'
```
Any reason why sortable isn't true for Service name? I'm playing around with this PR and it seems weird that I can't sort by Service name but can for all other table headers.

formgeist replied:
Yeah, that column should be sortable. Also, I think we should default to sort by service.

ogupte replied:
Thanks for pointing that out! 😄

bmorelli25 reviewed (approved):
🎉Nice! Copy LGTM! One question below that's unrelated to words.

formgeist commented:
Found a minor design issue: The services selector doesn't sort the services alphabetically.

<img width="390" alt="Screenshot 2019-07-01 at 21 37 35" src="https://user-images.githubusercontent.com/4104278/60462155-d3699080-9c48-11e9-89bd-31b3df5fa34b.png">

ogupte commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,346 @@
+    selectedConfig ? selectedConfig.service.name : undefined
+  );
+  const [sampleRate, setSampleRate] = useState<number | string>(
+    selectedConfig ? parseFloat(selectedConfig.settings.sample_rate) : ''
```
that's a good point, there's no reason why can't render NaN as '' in the input field. i'll make this change.

ogupte committed:
- PR feedback and fixed errors from CI

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5168/)

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,270 @@
+            'xpack.apm.settings.cm.flyOut.sampleRateConfigurationInputHelpText',
+            {
+              defaultMessage:
+                'Choose a rate between 0.00 and 1.0. Default configuration is 1.0 (100% of traces).'
```
Should say `0.000` since we added another decimal to the validation for 0.1% sample rate.

formgeist replied:
```suggestion
                'Choose a rate between 0.000 and 1.0. Default configuration is 1.0 (100% of traces).'
```

formgeist commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,270 @@
+          error={i18n.translate(
+            'xpack.apm.settings.cm.flyOut.sampleRateConfigurationInputErrorText',
+            {
+              defaultMessage: 'Sample rate must be between 0.00 and 1'
```
```suggestion
              defaultMessage: 'Sample rate must be between 0.000 and 1'
```

formgeist reviewed (dismissed):
Just a minor fix for the help texts to clarify that you can set it to 0.001 if you want to sample 0.1%.

dgieselaar reviewed (approved):
LGTM, congrats on getting this in!

formgeist commented:
Grea job everybody!

![](https://media.giphy.com/media/wrzf9P70YWLJK/giphy-downsized.gif)

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,346 @@
+    return null;
+  }
+
+  const [environment, setEnvironment] = useState<string | undefined>(
```
It is pretty verbose - but I <3 simple :D
This article is a bit overused but I like linking to it nonetheless https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,344 @@
+}) {
+  try {
+    if (isNaN(sampleRate) || !serviceName) {
+      throw new Error('Missing arguments');
```
We might want to make this a little more user friendly. But that's probably a nice-to-have.

ogupte replied:
If this error gets thrown, then something is really wrong with the UI. with all the validation checks and disabling the save button, this should never really happen, unless someone is tinkering with the client code.

sqren replied:
Ah okay. Then maybe just remove the check. Or is it just there to please Typescript?

sqren commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/get_service_names.ts:
```diff
@@ -0,0 +1,47 @@
+  SERVICE_NAME
+} from '../../../../common/elasticsearch_fieldnames';
+
+export type CMServicesAPIResponse = PromiseReturnType<typeof getServiceNames>;
```
Terminology: When I started this PR I was unsure about what terminology we were using. Settings? Central config management? And how was it abbreviated? So I might have muddied the waters from the get go.
Would be nice to get this cleaned up a bit.

In this case I see `CMServicesAPIResponse` where "CM" is a (bad) abbreviation of Central Config Management. Since we now call it Agent Remote Configuration perhaps this abbreviation is not so apt anymore.

formgeist replied:
We actually landed on APM Agent Configuration as the feature name (https://github.com/elastic/apm-dev/issues/553) 🙂

sqren replied:
Great! I'll try to remember that and stop saying central configuration management :D

sorenlouv reviewed (commented):
Great job! Just added a few questions - if you have already gone over some of it you can just disregard my comment.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -0,0 +1,270 @@
+                (isNaN(sampleRate) || isSampleRateValid)) ||
+              (!selectedConfig &&
+                (!(serviceName || environment) ||
+                  (isNaN(sampleRate) || isSampleRateValid)))
```
Can this be extracted to something like `shouldShowSampleRateError`?

ogupte replied:
i plan on fixing some of the issues with validation in another issue (https://github.com/elastic/kibana/issues/40114) so i can clean up the ui validation code there as well

sqren commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/create_configuration.ts:
```diff
@@ -0,0 +1,36 @@
+    index: config.get<string>('apm_oss.cmIndex'),
+    body: {
+      timestamp: {
+        us: Date.now() * 1000
```
In the index creation `@timestamp` is used but here it is `timestamp.us`

ogupte replied:
https://github.com/elastic/kibana/pull/39555#discussion_r299577641

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/SettingsList.tsx:
```diff
@@ -0,0 +1,297 @@
+    }
+  );
+
+  const hasConfigurations = data.length !== 0;
```
Could `data.length` conceivably be `undefined`? In that case `data.length > 0` would be better. Or `!isEmpty(data)` which is more readable imo.

ogupte commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/create_cm_index.ts:
```diff
@@ -0,0 +1,54 @@
+          properties: {
+            '@timestamp': {
+              type: 'date'
+            },
```
Good point. The original issue describes a documents with this example:
```
{
  "_id": "s4uMZJqvfrNo7sa",
  "settings": {
    "sample_rate": 0.5
  },
  "@timestamp": "2019-04-12T14:00:17.700Z",
  "service": {
    "name": "opbeans-node",
    "environment": "prod"
  }
}
```
So based on this i used type `date` for `@timestamp`, but then the existing code that i started with was setting and reading from `timestamp.us`, so by writing to that field, it automatically updates the mapping to make the new field valid.

```
"timestamp" : {
    "properties" : {
    "us" : {
        "type" : "long"
    }
    }
}
```

I'm not sure the best way to proceed here. it looks like we use both `@timestamp` (date) and `timestamp.us` (long) in the apm indices. But we don't need both here. We also don't really need the microseconds granularity since only the UI write and reads from, so it's always `Date.now()+1000` anyways and is never set to anything more specific.

My intuition tells me to go with `@timestamp` (date) with epoch time in milliseconds, since that's what the UI needs.

sqren replied:
Yeah, I'm fine with either as long as we stick to one.

ogupte commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/create_cm_index.ts:
```diff
@@ -0,0 +1,54 @@
+            settings: {
+              properties: {
+                sample_rate: {
+                  type: 'text'
```
yes, it was decided to have all fields be stored as text since everything ends up as text at the agent configuration level

ogupte commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/create_cm_index.ts:
```diff
@@ -0,0 +1,54 @@
+            service: {
+              properties: {
+                name: {
+                  type: 'keyword'
```
i honestly never considered setting `ignore_above`, can you think of a good value to set it to?

simitt replied:
APM Server side we set `ignore_above: 1024` for data indexed as `keyword`.

ogupte commented on x-pack/legacy/plugins/apm/server/lib/settings/cm/delete_configuration.ts:
```diff
@@ -0,0 +1,25 @@
+  const { client, config } = setup;
+
+  const params = {
+    refresh: true,
```
Haven't discussed it, but i can change it to `'wait_for'` so as to be unobtrusive for the rest of cluster. It adds an almost unnoticeable delay to delete, but i think it's actually OK, ux-wise since there's more gravity to the delete operation.

ogupte committed:
- changed all uses of timestamp.us to just '@timestamp' (epoch millis)
- memoized returned object from useFetcher when data doesn't change
- changed delete query refresh from true to 'wait_for'
- added ignore_above: 1024 to all keyword typed fields in .apm-cm index
- fixed some wording around sample_rate granularity
- various code cleanup

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5380/)

sqren commented on x-pack/legacy/plugins/apm/public/hooks/useFetcher.tsx:
```diff
@@ -25,47 +26,63 @@ export function useFetcher<Response>(
-            error: undefined
-          });
+      async function doFetch() {
+        const promise = fn();
```
I did the updates to `useFetcher` so I can probably explain :)
Most places `useFetcher` is called with a conditional promise returning function:
```js
  const { data } = useFetcher(
    () => {
      if (serviceName && start && end) {
        return loadErrorDistribution({ serviceName, start, end });
      }
    },
    [serviceName, start, end]
  );
```

It caused some weird loading indicator issues when `useFetcher` changed the state to LOADING, and then immediately changed it to SUCCESS if the condition was falsy (and thus the promise just resolved with `undefined`).

Now if the function doesn't return a promise `useFetcher` will not change the state to LOADING (and subsequently not to SUCCESS either).

LMK if that explanation makes sense. I'm open to other ways to get around this problem - this was just the first thing that came to my mind.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,344 @@
+  );
+  const { data: serviceNames = [], status: serviceNamesStatus } = useFetcher<
+    string[]
+  >(async () => (await loadCMServices()).sort(), [], {
```
You don't need to supply a type to `useFetcher`. It is inferred from the promise return type.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,344 @@
+    selectedConfig ? parseFloat(selectedConfig.settings.sample_rate) : NaN
+  );
+  const { data: serviceNames = [], status: serviceNamesStatus } = useFetcher<
+    string[]
```
You don't need to supply a type to `useFetcher`. It is inferred from the promise return type.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,344 @@
+    string[]
+  >(async () => (await loadCMServices()).sort(), [], {
+    preservePreviousResponse: false
+  });
```
I think you should move the `sort` to the rest method (`loadCMServices`) and jsut do:
```js
useFetcher(loadCMServices, [], { preservePreviousResponse: false });
```

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -0,0 +1,344 @@
+  });
+  const { data: environments = [], status: environmentStatus } = useFetcher<
+    Array<{ name: string; available: boolean }>
+  >(
```
Same as above: no type needs to be supplied to `useFetcher`
